src/webgpu/api/operation/memory_sync/buffer.spec.ts

```diff
@@ -0,0 +1,367 @@
+      })
+      .createView();
+    return encoder.beginRenderPass({
+      colorAttachments: [
+        {
+          attachment: view,
+          loadValue: { r: 0.0, g: 1.0, b: 0.0, a: 1.0 },
+        },
+      ],
+    });
+  }
+
+  // Write buffer via draw call in render pass. Use bundle if needed.
+  writeByRenderPass(bundle: boolean, buffer: GPUBuffer, value: number, encoder: GPUCommandEncoder) {
+    const pipeline = this.createStorageWriteRenderPipeline(value);
+    const bindGroup = this.createBindGroup(pipeline, buffer);
+
+    const pass = this.beginSimpleRenderPass(encoder);
+    const renderer = bundle
+      ? this.device.createRenderBundleEncoder({ colorFormats: ['rgba8unorm'] })
+      : pass;
+    renderer.setBindGroup(0, bindGroup);
+    renderer.setPipeline(pipeline);
+    renderer.draw(1, 1, 0, 0);
+
+    if (bundle) pass.executeBundles([(renderer as GPURenderBundleEncoder).finish()]);
+    pass.endPass();
+  }
+
+  // Write buffer via dispatch call in compute pass.
+  writeByComputePass(buffer: GPUBuffer, value: number, encoder: GPUCommandEncoder) {
+    const pipeline = this.createStorageWriteComputePipeline(value);
+    const bindGroup = this.createBindGroup(pipeline, buffer);
+    const pass = encoder.beginComputePass();
+    pass.setPipeline(pipeline);
+    pass.setBindGroup(0, bindGroup);
+    pass.dispatch(1);
+    pass.endPass();
+  }
+
+  // Write bufer via BuferToBuffer copy.
+  writeByB2BCopy(buffer: GPUBuffer, value: number, encoder: GPUCommandEncoder) {
+    const tmpBuffer = this.createBufferWithValue(value);
+    encoder.copyBufferToBuffer(tmpBuffer, 0, buffer, 0, SIZE);
+  }
+
+  // Write buffer via TextureToBuffer copy.
+  writeByT2BCopy(buffer: GPUBuffer, value: number, encoder: GPUCommandEncoder) {
+    const tmpBuffer = this.createBufferWithValue(value);
+    const tmpTexture = this.device.createTexture({
+      size: { width: 1, height: 1, depth: 1 },
+      format: 'rgba8uint',
+      usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.COPY_DST,
+    });
+    encoder.copyBufferToTexture(
+      { buffer: tmpBuffer, bytesPerRow: 256 },
+      { texture: tmpTexture, mipLevel: 0, origin: { x: 0, y: 0, z: 0 } },
+      { width: 1, height: 1, depth: 1 }
+    );
+    encoder.copyTextureToBuffer(
+      { texture: tmpTexture, mipLevel: 0, origin: { x: 0, y: 0, z: 0 } },
+      { buffer, bytesPerRow: 256 },
+      { width: 1, height: 1, depth: 1 }
+    );
+  }
+
+  // Write buffer via writeBuffer API on queue
+  writeByWriteBuffer(buffer: GPUBuffer, value: number) {
+    const data = new Uint32Array(SIZE / 4);
+    for (let i = 0; i < SIZE / 4; ++i) {
+      data[i] = value;
+    }
+
+    this.device.defaultQueue.writeBuffer(buffer, 0, data);
+  }
+
+  // Write buffer via mapping and writing
+  writeByMapWrite(buffer: GPUBuffer, value: number) {
+    const data = new Uint32Array(SIZE / 4);
+    for (let i = 0; i < SIZE / 4; ++i) {
+      data[i] = value;
+    }
+
+    buffer.mapAsync(GPUMapMode.WRITE);
```

Comment:
mapSync is async, looks like we need an `await` or stuff like that? But I can't use `await` here because this function is not async. Any idea about this?

Reply:
Why can't you make writeByMapWrite async (and so on up the call stack)?

Reply:
Done. But the tests still fail and complained that "The buffer must have the MapWrite usage" when I run the tests in Chromium. anything wrong in tests?

Reply:
The buffer is not created with the MAP_WRITE bit. `mappedAtCreation` buffers cannot be remapped after creation unless they have that bit.

Reply:
If I understand correctly, MAP_WRITE usage can be compatible with COPY_SRC only. This means that MapWrite operation is impossible to be included into write_after_write test? Because another write operation is either a storage buffer or a COPY_DST buffer, which is not compatible with MAP_WRITE usage.

Reply:
removed mapWrite writeOperation related tests for write-after-write tests. We can add it when we test read-after-write and read-before-write.

Reply:
Right. Sorry I didn't catch that. We may not want do the map synchronization tests in this file, since it would only be tested in conjunction with copies.

That said, if it were tightly abstracted along with the rest of the ways of writing to and reading from buffers, it would probably be easiest to do it here (would just have to filter out the test cases that wouldn't be valid).

Reply:
Yeah. I intended to do so, too. MapWrite can be tested in read-after-write or read-before-write. And it might be in a separate tests (not combined with others) because it can be tested with copy only for the sake of buffer usage compatibility.